syncer/db.go

```diff
@@ -62,229 +71,171 @@ type binlogSize struct {
-			tctx.L().Warn("execute statements", zap.Int("retry", i), zap.Strings("sqls", sqls), zap.Reflect("arguments", args))
-			time.Sleep(retryTimeout)
-		}
+func (conn *Conn) getParser(ansiQuotesMode bool) (*parser.Parser, error) {
+	return utils.GetParser(conn.baseConn.DB, ansiQuotesMode)
+}
 
-		if err = conn.executeSQLImp(tctx, sqls, args); err != nil {
-			if isRetryableError(err) {
-				continue
-			}
-			tctx.L().Error("execute statements", zap.Strings("sqls", sqls), zap.Reflect("arguments", args), log.ShortError(err))
-			return terror.DBErrorAdapt(err, terror.ErrDBExecuteFailed, strings.Join(sqls, ";"))
-		}
+func (conn *Conn) killConn(connID uint32) error {
+	return utils.KillConn(conn.baseConn.DB, connID)
+}
 
-		return nil
-	}
+func (conn *Conn) fetchAllDoTables(bw *filter.Filter) (map[string][]string, error) {
+	return utils.FetchAllDoTables(conn.baseConn.DB, bw)
+}
 
-	tctx.L().Error("exec statements failed", zap.Strings("sqls", sqls), zap.Reflect("arguments", args), log.ShortError(err))
-	return terror.DBErrorAdapt(err, terror.ErrDBExecuteFailed, strings.Join(sqls, ";"))
+func (conn *Conn) countBinaryLogsSize(pos mysql.Position) (int64, error) {
+	return countBinaryLogsSize(pos, conn.baseConn.DB)
 }
 
-// Note: keep it for later use?
-func (conn *Conn) executeSQLImp(tctx *tcontext.Context, sqls []string, args [][]interface{}) error {
-	if conn == nil || conn.db == nil {
-		return terror.ErrDBUnExpect.Generate("database connection not valid")
+func (conn *Conn) querySQL(tctx *tcontext.Context, query string, args ...interface{}) (*sql.Rows, error) {
+	if conn == nil || conn.baseConn == nil {
+		return nil, terror.ErrDBUnExpect.Generate("database base connection not valid")
+	}
+	params := retry.Params{
+		RetryCount:         10,
+		FirstRetryDuration: retryTimeout,
+		BackoffStrategy:    retry.Stable,
+		IsRetryableFn: func(retryTime int, err error) bool {
+			if retry.IsRetryableError(err) {
+				tctx.L().Warn("query statement", zap.Int("retry", retryTime), zap.String("query", query), zap.Reflect("argument", args))
+				sqlRetriesTotal.WithLabelValues("query", conn.cfg.Name).Add(1)
+				return true
+			}
+			return false
+		},
 	}
 
-	startTime := time.Now()
-	defer func() {
-		txnHistogram.WithLabelValues(conn.cfg.Name).Observe(time.Since(startTime).Seconds())
-	}()
+	ret, _, err := conn.baseConn.ApplyRetryStrategy(
+		tctx,
+		params,
+		func(ctx *tcontext.Context) (interface{}, error) {
+			return conn.baseConn.QuerySQL(ctx, query, args...)
+		},
+	)
 
-	txn, err := conn.db.Begin()
 	if err != nil {
-		tctx.L().Error("begin transaction", zap.Strings("sqls", sqls), zap.Reflect("arguments", args), zap.Error(err))
-		return terror.DBErrorAdapt(err, terror.ErrDBExecuteFailed, strings.Join(sqls, ";"))
+		tctx.L().Error("query statement failed after retry", zap.String("query", query), zap.Reflect("argument", args), log.ShortError(err))
+		return nil, terror.ErrDBQueryFailed.Delegate(err, query)
```

Comment:
`err` is already a `*terror.Error` object, we should return `err` instead of using `Delegate`.

Reply:
ok, Done